File: src/IPlot.HighCharts/HighChartsChart.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;
using Microsoft.FSharp.Core;
using Newtonsoft.Json;

namespace IPlot.HighCharts
{
    /// The HighCharts chart object, containing all chart properties and traces
    public class HighChartsChart : ChartElement
    {
        /// Constructor with default chart property initialisation
        public HighChartsChart()
        {
            this.chart.chart_iplot = new Chart_iplot()
            {
                width = 900,
                height = 500
            };
            this.chart.title = new Title()
            {
                text = ""
            };
        }

        /// The chart object containing all chart properties
        public HighChart chart { get; set; } = new HighChart();

        /// The chart's container div id
        public string id { get; set; } = Guid.NewGuid().ToString();

        /// The highcharts.js src
        public string highchartsSrc { get; set; } = Html.DefaultHighChartsSrc;

        /// The collection of trace labels
        private IEnumerable<string> _labels;

        /// Set a specific chart property, using the Chart.Props helper (F# style)
        public static FSharpFunc<Func<HighChartsChart, HighChartsChart>, FSharpFunc<HighChartsChart, HighChartsChart>> WithFs
        {
            get
            {
                Func<Func<HighChartsChart, HighChartsChart>, HighChartsChart, HighChartsChart> lam = (propFun, HighChartsChart) => propFun((HighChartsChart)HighChartsChart.DeepClone());
                var f = FuncConvert.FromFunc<Func<HighChartsChart, HighChartsChart>, HighChartsChart, HighChartsChart>(lam);

                return f;
            }
        }

        /// Set a specific chart property, using the Chart.Props helper
        public static HighChartsChart With(Func<HighChartsChart, HighChartsChart> propFun, HighChartsChart HighChartsChart)
        {
            return propFun((HighChartsChart)HighChartsChart.DeepClone());
        }

        /// Set a specific chart property, using the Chart.Props helper
        public HighChartsChart With(Func<HighChartsChart, HighChartsChart> propFun)
        {
            return With(propFun, this);
        }

        /// Deep clone of the chart and all its properties
        public override ChartElement DeepClone()
        {
            var highchartsChart = new HighChartsChart();
            highchartsChart.chart = (HighChart)this.chart.DeepClone();
            highchartsChart.id = this.id;

            return highchartsChart;
        }

        /// Inline markup that can be embedded in a HTML document
        public string GetInlineHtml()
        {
            var plotting = GetPlottingJS();
            return
                Html.inlineTemplate
                    .Replace("[ID]", this.id)
                    .Replace("[WIDTH]", this.chart.chart_iplot.width.ToString())
                    .Replace("[HEIGHT]", this.chart.chart_iplot.height.ToString())
                    .Replace("[PLOTTING]", plotting);
        }

        /// Inline javascript that can be embedded in a HTML document
        public string GetInlineJS()
        {
            var plotting = GetPlottingJS();
            return
                Html.jsTemplate
                    .Replace("[PLOTTING]", plotting);
        }

        /// Serialise chart object to JSON
        public string SerializeChart()
        {
            return JsonConvert.SerializeObject(this.chart, Formatting.None, new JsonSerializerSettings
                {
                    NullValueHandling = NullValueHandling.Ignore
                })
                .Replace("_iplot", string.Empty)
                .Replace("data_mat", "data")
                .Replace("data_obj", "data");
        }

        /// Returns the JS to load relevant HighCharts modules
        public static IEnumerable<string> GetModules(HighChartsChart chart)
        {
            var seriesTypes = chart.chart.series.Select(s => s.type_iplot);
            var modules =
                seriesTypes
                .SelectMany(t => ModuleLoading.GetDependencies(t))
                .Distinct()
                .ToArray();

            var moduleImports =
                modules
                    .Select(m =>
                        Html.moduleTemplate
                            .Replace("[MODULENAME]", m)
                            .Replace("[HIGHCHARTSSRC]", chart.highchartsSrc));

            return moduleImports;
        }

        /// Returns the chart's full HTML source.
        private string GetHtml()
        {
            var chartMarkup = GetInlineHtml();
            var moduleSrc = String.Join("\n", GetModules(this));

            return
                Html.pageTemplate
                    .Replace("[CHART]", chartMarkup)
                    .Replace("[MODULESRC]", moduleSrc)
                    .Replace("[HIGHCHARTSSRC]", highchartsSrc);
        }

        /// The chart's plotting JavaScript code.
        private string GetPlottingJS()
        {
            var chartJson = SerializeChart();

            return
                Html.jsFunctionTemplate
                    .Replace("[THEME]", Html.themeString)
                    .Replace("[ID]", this.id)
                    .Replace("[CHARTOBJ]", chartJson);
        }

        /// Generate a chart from a collection of traces
        public void Plot(IEnumerable<Trace> data)
        {
            this.chart.series = data.ToList();
        }

        /// Display the chart in a browser
        public void Show()
        {
            var html = GetHtml();
            Html.showInBrowser(html, this.id);
        }

        /// Combine charts together and display as a single page in default browser
        public static void ShowAll(IEnumerable<HighChartsChart> charts)
        {
            var html = string.Join("",charts.Select(c => c.GetInlineHtml()));
            var highChartsSrc = charts.Any() ? charts.First().highchartsSrc : Html.DefaultHighChartsSrc;
            var modules =
                charts
                .SelectMany(c => GetModules(c))
                .Distinct()
                .ToArray();
            var moduleSrc = String.Join("\n", modules);

            var pageHtml =
                Html.pageTemplate
                .Replace("[CHART]", html)
                .Replace("[MODULESRC]", moduleSrc)
                .Replace("[HIGHCHARTSSRC]", highChartsSrc);

            var combinedChartId = Guid.NewGuid().ToString();            
            Html.showInBrowser(pageHtml, combinedChartId);
        }

        /// Save the chart to an HTML file
        public void SaveHtml(string path)
        {
            var html = GetHtml();
            var dir = Path.GetDirectoryName(path);

            if (!String.IsNullOrEmpty(dir))
            {
                if (!Directory.Exists(dir))
                    Directory.CreateDirectory(dir);
            }

            File.WriteAllText(path, html);
        }

        /// Combine charts together and save as a single HTML file
        public static void SaveHtmlAll(string path, IEnumerable<HighChartsChart> charts)
        {
            var html = string.Join("",charts.Select(c => c.GetInlineHtml()));
            var highChartsSrc = charts.Any() ? charts.First().highchartsSrc : Html.DefaultHighChartsSrc;
            var modules =
                charts
                .SelectMany(c => GetModules(c))
                .Distinct();
            var moduleSrc = String.Join("\n", modules);

            var pageHtml =
                Html.pageTemplate
                .Replace("[CHART]", html)
                .Replace("[MODULESRC]", moduleSrc)
                .Replace("[HIGHCHARTSSRC]", highChartsSrc);           

            var dir = Path.GetDirectoryName(path);

            if (!String.IsNullOrEmpty(dir))
            {
                if (!Directory.Exists(dir))
                    Directory.CreateDirectory(dir);
            }

            File.WriteAllText(path, pageHtml);
        }

        /// Sets the chart's title
        public HighChartsChart WithTitle(string title)
        {
            this.chart.title.text = title;
            return this;
        }

        /// Sets the chart's X-axis title
        public HighChartsChart WithXTitle(string xTitle)
        {
            if (this.chart.xAxis == null)
                this.chart.xAxis = new XAxis[] { new XAxis() { title = new Title() { text = xTitle } } };
            else if (!this.chart.xAxis.Any())
                this.chart.xAxis = new XAxis[] { new XAxis() { title = new Title() { text = xTitle } } };
            else
            {
                if (this.chart.xAxis.First().title == null)
                    this.chart.xAxis.First().title = new Title() { text = xTitle };
                else
                    this.chart.xAxis.First().title.text = xTitle;
            }

            return this;
        }

        /// Sets the chart's Y-axis title
        public HighChartsChart WithYTitle(string yTitle)
        {
            if (this.chart.yAxis == null)
                this.chart.yAxis = new YAxis[] { new YAxis() { title = new Title() { text = yTitle } } };
            else if (!this.chart.yAxis.Any())
                this.chart.yAxis = new YAxis[] { new YAxis() { title = new Title() { text = yTitle } } };
            else
            {
                if (this.chart.yAxis.First().title == null)
                    this.chart.yAxis.First().title = new Title() { text = yTitle };
                else
                    this.chart.yAxis.First().title.text = yTitle;
            }

            return this;
        }

        /// Sets the chart's width
        public HighChartsChart WithWidth(int width)
        {
            this.chart.chart_iplot.width = width;
            return this;
        }

        /// Sets the chart's height
        public HighChartsChart WithHeight(int height)
        {
            this.chart.chart_iplot.height = height;
            return this;
        }

        /// Sets the chart's container div id
        public HighChartsChart WithId(string id)
        {
            this.id = id;
            return this;
        }

        /// Sets the data series label. Use this member if the
        /// chart's data is a single series
        public HighChartsChart WithLabel(string label)
        {
            _labels = new string[] { label };
            return this;
        }

        /// Sets the data series labels. Use this method if the
        /// chart's data is a series collection
        public HighChartsChart WithLabels(IEnumerable<string> labels)
        {
            _labels = labels.ToArray();
            return this;
        }

        /// Sets the chart's width and height
        public HighChartsChart WithSize(int width, int height)
        {
            WithWidth(width);
            WithHeight(height);
            return this;
        }
    }
}

File: src/IPlot/HighCharts.Chart.fs
﻿namespace IPlot.HighCharts

open System
open System.Net.Http
open System.Text
open System.Net
open System.IO

type key = IConvertible
type value = IConvertible

type Chart() =
    static member val Props = HighChart_IProp() with get

    static member internal DateTimeToHighCharts (dt:DateTime) =
        dt.Subtract(DateTime(1970,1,1,0,0,0,DateTimeKind.Utc)).TotalMilliseconds
    
    static member With (propFun: Func<HighChartsChart, HighChartsChart>) (chart: HighChartsChart) =
        propFun.Invoke (chart.DeepClone() :?> HighChartsChart)

    static member Plot data =
        let chart = HighChartsChart()
        chart.Plot [data]
        chart

    static member Plot (data: seq<#Trace>) =
        let chart = HighChartsChart()
        chart.Plot (Seq.cast<Trace> data)
        chart
    
    /// Displays a chart in the default browser
    static member Show (chart: HighChartsChart) = chart.Show()
    
    /// Combine charts together and display as a single page in default browser
    static member ShowAll(charts: seq<HighChartsChart>) = HighChartsChart.ShowAll charts

    static member Save (filename:string) (chart: HighChartsChart) =
        let json = chart.SerializeChart()

        try
            use clientHandler = new HttpClientHandler()
            clientHandler.ServerCertificateCustomValidationCallback <- (fun sender cert chain sslPolicyErrors -> true)

            use client = new HttpClient(clientHandler)
            client.Timeout <- TimeSpan.FromSeconds(3.)

            let jsonTemplate = "{
                \"options\":##JSON##,
                \"type\":\"image/png\",
                \"async\":false
            }"

            let contentStr = jsonTemplate.Replace("##JSON##", json)
            use content = new StringContent(contentStr, Encoding.UTF8, "application/json")

            let resp =
                client.PostAsync("https://export.highcharts.com/", content)
                |> Async.AwaitTask
                |> Async.RunSynchronously
            
            match resp.StatusCode with
            | HttpStatusCode.OK ->
                let content =
                    resp.Content.ReadAsStreamAsync()
                    |> Async.AwaitTask
                    |> Async.RunSynchronously
                
                let p = Path.GetDirectoryName filename
                if String.IsNullOrEmpty p |> not && Directory.Exists p |> not then
                    Directory.CreateDirectory p
                    |> ignore
                                
                use fs = File.Create(filename)
                content.Seek(0L, SeekOrigin.Begin) |> ignore
                content.CopyTo(fs)
                printfn "Written %i bytes to %s" content.Length filename
            | _ ->
                printfn "Failed to save chart: %s" (string resp.StatusCode)
        with ex ->
            printfn "Exception saving chart: %s" (string ex)
    
    /// Displays a chart in the default browser
    static member SaveHtml (path: string) (chart: HighChartsChart) = chart.SaveHtml path
    
    /// Combine charts together and display as a single page in default browser
    static member SaveHtmlAll (path: string) (charts: seq<HighChartsChart>) = HighChartsChart.SaveHtmlAll(path, charts)

    /// Sets the chart's width.
    static member WithTitle title (chart:HighChartsChart) =
        chart.WithTitle title

    /// Sets the chart's X-axis title.
    static member WithXTitle xTitle (chart:HighChartsChart) =
        chart.WithXTitle xTitle

    /// Sets the chart's Y-axis title.
    static member WithYTitle yTitle (chart:HighChartsChart) =
        chart.WithYTitle yTitle

    /// Sets the chart's width.
    static member WithWidth width (chart:HighChartsChart) =
        chart.WithWidth width
    
    /// Sets the chart's height.
    static member WithHeight height (chart:HighChartsChart) =
        chart.WithHeight height

    /// Sets the chart's container div id.
    static member WithId id (chart:HighChartsChart) =
        chart.WithId id

    /// Sets the data series label. Use this member if the
    /// chart's data is a single series.
    static member WithLabel label (chart:HighChartsChart) =
        chart.WithLabel label

    /// Sets the data series labels. Use this member if the
    /// chart's data is a series collection.
    static member WithLabels labels (chart:HighChartsChart) =
        chart.WithLabels labels

    /// Sets the chart's height.
    static member WithSize size (chart:HighChartsChart) =
        chart.WithSize size

    static member internal ToFloatArray s =
        s
        |> Seq.map (fun v -> (v :> IConvertible).ToDouble(null))
        |> Seq.toArray

    static member internal ToFloatArray2d s =
        s
        |> Seq.map (fun (x,y) -> seq {(x :> IConvertible).ToDouble(null);(y :> IConvertible).ToDouble(null)})
        |> Seq.toArray

    static member internal ToFloatArray2d (s:(float*float) seq) =
        s
        |> Seq.map (fun (x,y) -> seq { x; y })
        |> Seq.toArray

    static member internal ToTimeFloatArray s =
        s
        |> Seq.map (fun (x:DateTime,y:float) -> seq { Chart.DateTimeToHighCharts x; y })
        |> Seq.toArray

    static member internal ToStringArray s =
        s
        |> Seq.map (fun s -> s.ToString())
        |> Seq.toArray

    static member Cylinder(data:seq<#value>) =
        let cylinder = Cylinder(data = (data |> Chart.ToFloatArray))
        Chart.Plot [cylinder]
        |> Chart.With (Chart.Props.chart_iplot.options3d.enabled true)
        |> Chart.With (Chart.Props.chart_iplot.options3d.alpha 15.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.beta 15.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.depth 20.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.viewDistance 500.)

    static member Funnel(data:seq<#value>) =
        let funnel = Funnel(data = (data |> Chart.ToFloatArray))
        Chart.Plot [funnel]

    static member Funnel3d(data:seq<#value>) =
        let funnel = Funnel3d(data = (data |> Chart.ToFloatArray))
        Chart.Plot [funnel]
        |> Chart.With (Chart.Props.chart_iplot.options3d.enabled true)
        |> Chart.With (Chart.Props.chart_iplot.options3d.alpha 10.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.beta 100.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.depth 200.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.viewDistance 50.)    

    static member Line(data:seq<#value>) =
        let scatter =
            Line(data = (data |> Chart.ToFloatArray),
                marker = Marker(enabled = Nullable<bool>(false)))
        Chart.Plot [scatter]

    static member Line(data:seq<float>) =
        let scatter =
            Line(data = data,
                marker = Marker(enabled = Nullable<bool>(false)))
        Chart.Plot [scatter]

    static member Line(data:seq<#key * #value>) =
        let scatter = 
            Line(
                data_mat = Chart.ToFloatArray2d data,
                marker = Marker(enabled = Nullable<bool>(false)))
        Chart.Plot [scatter]

    static member Line(data:seq<float * float>) =
        let scatter =
            Line(
                data_mat = Chart.ToFloatArray2d data,
                marker = Marker(enabled = Nullable<bool>(false)))
        Chart.Plot [scatter]

    static member Line(data:seq<DateTime * float>) =
        let scatter =
            Line(
                data_mat = Chart.ToTimeFloatArray data,
                marker = Marker(enabled = Nullable<bool>(false)))
        Chart.Plot [scatter]
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Line(data:seq<seq<#value>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data = (series |> Chart.ToFloatArray),
                    marker = Marker(enabled = Nullable<bool>(false))
                ))
        Chart.Plot scatters

    static member Line(data:seq<seq<float>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data = series,
                    marker = Marker(enabled = Nullable<bool>(false))
                ))
        Chart.Plot scatters

    static member Line(data:seq<float []>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data = series,
                    marker = Marker(enabled = Nullable<bool>(false))
                ))
        Chart.Plot scatters

    static member Line(data:seq<float list>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data = series,
                    marker = Marker(enabled = Nullable<bool>(false))
                ))
        Chart.Plot scatters

    static member Line(data:seq<#seq<#key * #value>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = Chart.ToFloatArray2d series,
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters

    static member Line(data:seq<seq<float * float>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = (series |> Seq.map (fun (x,y) -> seq { x; y })),
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters

    static member Line(data:seq<(float * float) []>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = (series |> Seq.map (fun (x,y) -> seq { x; y })),
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters

    static member Line(data:seq<(float * float) list>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = (series |> Seq.map (fun (x,y) -> seq { x; y })),
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters

    static member Line(data:seq<seq<DateTime * float>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = (series |> Seq.map (fun (x,y) -> seq { Chart.DateTimeToHighCharts x; y })),
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Line(data:seq<(DateTime * float) []>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = (series |> Array.map (fun (x,y) -> seq { Chart.DateTimeToHighCharts x; y })),
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Line(data:seq<(DateTime * float) list>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Line(
                    data_mat = (series |> List.map (fun (x,y) -> seq { Chart.DateTimeToHighCharts x; y })),
                    marker = Marker(enabled = Nullable<bool>(false))))
        Chart.Plot scatters
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Pyramid3d(data:seq<#value>) =
        let pyramid = Pyramid3d(data = (data |> Chart.ToFloatArray))
        Chart.Plot [pyramid :> Trace]
        |> Chart.With (Chart.Props.chart_iplot.options3d.enabled true)
        |> Chart.With (Chart.Props.chart_iplot.options3d.alpha 10.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.beta 100.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.depth 200.)
        |> Chart.With (Chart.Props.chart_iplot.options3d.viewDistance 50.)

    static member Scatter(data:seq<#value>) =
        let scatter =
            Scatter(data = (data |> Chart.ToFloatArray))
        Chart.Plot [scatter]

    static member Scatter(data:seq<float>) =
        let scatter = Scatter(data = data)
        Chart.Plot [scatter]

    static member Scatter(data:seq<float * float>) =
        let scatter = Scatter(data_mat = Chart.ToFloatArray2d data)
        Chart.Plot [scatter]

    static member Scatter(data:seq<DateTime * float>) =
        let scatter = Scatter(data_mat = Chart.ToTimeFloatArray data)
        Chart.Plot [scatter]
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Scatter(data:seq<seq<#value>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(data = (series |> Chart.ToFloatArray)))
        Chart.Plot scatters

    static member Scatter(data:seq<seq<float>>) =
        let scatter = Scatter(data_mat = data)
        Chart.Plot [scatter]

    static member Scatter(data:seq<float []>) =
        let scatter = Scatter(data_mat = (data |> Seq.map (Seq.ofArray)))
        Chart.Plot [scatter]

    static member Scatter(data:seq<float list>) =
        let scatter = Scatter(data_mat = (data |> Seq.map (Seq.ofList)))
        Chart.Plot [scatter]

    static member Scatter(data:seq<#seq<#key * #value>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(data_mat = Chart.ToFloatArray2d series))
        Chart.Plot scatters

    static member Scatter(data:seq<seq<float * float>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(data_mat = (series |> Seq.map (fun (x,y) -> seq { x; y }))))
        Chart.Plot scatters

    static member Scatter(data:seq<(float * float) []>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(data_mat = (series |> Array.map (fun (x,y) -> seq { x; y }))))
        Chart.Plot scatters

    static member Scatter(data:seq<(float * float) list>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(data_mat = (series |> List.map (fun (x,y) -> seq { x; y }))))
        Chart.Plot scatters

    static member Scatter(data:seq<seq<DateTime * float>>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(
                    data_mat = (series |> Seq.map (fun (x,y) -> seq { Chart.DateTimeToHighCharts x; y }))
                ))
        Chart.Plot scatters
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Scatter(data:seq<(DateTime * float) []>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(
                    data_mat = (series |> Array.map (fun (x,y) -> seq { Chart.DateTimeToHighCharts x; y }))
                ))
        Chart.Plot scatters
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")

    static member Scatter(data:seq<(DateTime * float) list>) =
        let scatters =
            data
            |> Seq.map (fun series ->
                Scatter(
                    data_mat = (series |> List.map (fun (x,y) -> seq { Chart.DateTimeToHighCharts x; y }))
                ))
        Chart.Plot scatters
        |> Chart.With (Chart.Props.xAxis.[0].type_iplot "datetime")
        
    // static member Area
    // static member Arearange
    // static member Areaspline
    // static member Areasplinerange
    // static member Bar
    // static member Bellcurve
    // static member Boxplot
    // static member Bubble
    // static member Bullet
    // static member Column
    // static member Columnpyramid
    // static member Columnrange
    // static member Dependencywheel
    // static member Dumbbell
    // static member Errorbar
    // static member Gauge
    // static member Heatmap
    // static member Histogram
    // static member Item
    // static member Line
    // static member Lollipop
    // static member Map
    // static member Mapbubble
    // static member Mapline
    // static member Mappoint
    // static member Networkgraph
    // static member Organization
    // static member Pareto
    // static member Pie
    // static member Polygon
    // static member Pyramid
    // static member Sankey
    // static member Scatter3d
    // static member Solidgauge
    // static member Spline
    // static member Streamgraph
    // static member Sunburst
    // static member Tilemap
    // static member Timeline
    // static member Treemap
    // static member Variablepie
    // static member Variwide
    // static member Vector
    // static member Venn
    // static member Waterfall
    // static member Windbarb
    // static member Wordcloud
    // static member Xrange
